testing/yield.js

```diff
@@ -0,0 +1,67 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Install "yield" support to Mocha tests.
+ * Check test-yield.js for how-to.
+ */
+export function installYieldIt(realIt) {
+  it = enableYield.bind(null, realIt); // eslint-disable-line no-native-reassign, no-undef
+  it./*OK*/only = enableYield.bind(null, realIt.only);
+  it.skip = realIt.skip;
+}
+
+function enableYield(fn, message, runnable) {
+  if (!runnable || !runnable.constructor
+      || runnable.constructor.name !== 'GeneratorFunction') {
+    return fn(message, runnable);
+  }
+  return fn(message, done => {
+    const runner = (iterator, result) => {
+      let state;
+      try {
+        state = iterator.next(result);
+      } catch (e) {
+        // catch any assertion errors and pass to `done`
+        // otherwise the messages are swallowed
+        return done(e);
+      }
+      if (state.done) {
+        return done();
+      }
+
+      const _runner = runner.bind(null, iterator);
+      if (isPromise(state.value)) {
+        // With this, we can do: `const result = yield promise;`
+        state.value.then(_runner).catch(done);
```

Comment:
It's not for error reporting, but so we can take full advantage of the generator's functionality. Imagine:

```js
it('regular code', function() {
  return somePromiseThatRejects.then(() => {
    throw new Error('UNREACHABLE');
  }, err => {
    expect(err.message).to.contain('something');
  });
});
```

As it is now, this cannot be written into our yield style, because the promise would reject and kill the test. Instead, we want something like:

```js
it('generator yield', function*() {
  try {
    yield somePromiseThatRejects;
    throw new Error('UNREACHABLE');
  } catch (e) {
    expect(e.message).to.contain('something');
  }
});
```

That's throwing the error into the iterator, allowing us to test all our promise tests as sync generators.